1. Line 12 will print out the final value of ```i``` at the end of the for loop, which is *3*. There isn't an error because we used ```var```, which is not block-scope. Also, the reason why ```i = 3``` is because the ```prices.length = 3```. By the 4th iteration of the for-loop (```i = 3```), ```i``` is no longer < ```prices.length```. So we will exit the for-loop with the final value of ```i``` being 3.
   
2. Line 13 will print out the final value of ```discountedPrice``` after the for-loop is done executing, which is *150*. There isn't any error because we also used ```var``` to declare ```discountedPrice```. Hence, we can use it even outside of the for-loop. The reason why ```discountedPrice = 150``` is because the last value of ```i``` that can still be used inside the for-loop is 2. We know that ```discountedPrice = prices[i] * (1 - discount)```. If we plug in all appropriate values, we will have ```discountedPrice = prices[2] * (1 - discount)```, which equals to ```300 * (1 - 0.5) = 150```.
   
3. Line 14 will print out the final value of ```finalPrice``` after the for-loop is done executing, which is *150*. There isn't any error because we also used ```var``` to declare ```finalPrice```. The reason why ```finalPrice = 150``` is because the final value of ```discountedPrice``` is 150. We know that ```finalPrice = Math.round(discountedPrice * 100)/100;```. If we plug in all appropriate values, we will have ```finalPrice = Math.round(150 * 100)/100 = 15000/100 = 150```.

4. ```discountPrices([100, 200, 300], .5)``` will return ```[50, 100, 150]``` because we are looping through the original array of prices, reducing the value of each item by half then add these new discounted prices into the new array call ```discounted```. This is what the function returns. Also, since we declared ```discounted``` with the ```var``` keyword, we can still access its final value even outside of the for-loop.

5. Line 12 gives me an error: ```ReferenceError: i is not defined```. This makes sense because we using the keyword ```let``` to declare ```i```. So ```i``` will only have the block-scope within the for-loop. Since line 12 is outside of the for-loop, using ```i``` will cause an error.

6. Line 13 will also give me an error: ```ReferenceError: discountedPrice is not defined```. The reason is similar to that in question 5: we are using the ```let``` keyword to declare ```discountedPrice``` inside the for-loop. Hence, using that variable in line 13, which is outside of the for-loop, will cause an error.

7. Line 14 will print out *150*, which is the final value of ```finalPrice```. The reason why this line doesn't throw an error despite using ```let``` is that we are declaring ```finalPrice``` outside of the for-loop and inside the general function. Thus, we can still access it before, inside and after the for-loop (which is inside the general function).